---
title: "Retrofit Colon Vignette"
author: "Adam Keebum Park, Roopali Singh"
date: "`r Sys.Date()`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Retrofit Colon Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
# knitr::opts_chunk$set(echo = TRUE, results = 'hide', fig.width = 4.5, fig.height = 4.5*(456/480))
knitr::opts_chunk$set(echo = TRUE, results = 'hide')
```

# Introduction

RETROFIT is a statistical method for reference-free deconvolution of spatial transcriptomics data to estimate cell type mixtures. In this Vignette, we will estimate cell type composition of a Colon dataset. We will annotate cell types using marker gene lists as well as single cell reference. 
We will also reproduce some of the results from the paper for illustration.

# Package Installation and other requirements

Install and load the packages using the following steps:
```{r, eval=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("retrofit")
```

```{r, load_library}
library(retrofit)
```

# Spatial Transcriptomics Data

First load the Colon ST data from fetal 12 PCW tissue, using the following command:

```{r, data}
data("vignetteColonData")
X = vignetteColonData$colon_a3_x
```

The data used here was pre-processed, similar to STdeconvolve, which includes the following:

## Xi: may need to add the pre-processing steps done by Xi using STdeconvolve package 

- remove spots with too few genes
- remove genes that are present in 100% of the spots
- remove genes that are present in less than 5% of the spots
- find over dispersed genes for deconvolution.

In the end, the ST data matrix will consist of G = 722 genes and S = 1080 spots i.e., a matrix of order G x S.

# Reference-free Deconvolution

Initialize the required parameters for deconvolution. We set the number of iterations, Iter = 4000 and number of components, L = 20. Lamda is kept at the default value of 0.01. 

After initialization, run retrofit on the data (X) as follows:

```{r, decompose}
Iter = 20
L = 20
result = retrofit::decompose(X, L=L, iterations=Iter, verbose=TRUE)

H = result$h
W = result$w
TH =result$th
```

After deconvolution of ST data, we have our estimates of W (a matrix of order G x L), H (a matrix of order L x S) and Theta (a vector of L components). Here, we are using number of iterations as 20 for demonstration purposes. For reproducing results in the paper, we need to run RETROFIT for iterations = 4000. The whole computation is omitted here due to time complexity (> 10min). We will load the results from 4000 iterations for the rest of the analysis.
```{r, load_retrofit_results}
H           = vignetteColonData$results_4k_iterations$decompose$h
W           = vignetteColonData$results_4k_iterations$decompose$w
Theta       = vignetteColonData$results_4k_iterations$decompose$th
################################################################################
# The above results are obtained by running the code below.
# iterations = 4000
# set.seed(1)
# result      = retrofit::decompose(x, L=L, iterations=iterations, verbose=TRUE)
################################################################################
```

Next, we need to annotate the components, to get the proportion of K cell types. We can do this in two ways: (a) using an annotated single cell reference or (b) using the known marker genes. 

# Cell-type Annotation via annotated single-cell reference

Here, we will annotate components using single-cell reference. Load the single-cell reference data:

```{r, sc}
sc_ref = vignetteColonData$sc_ref
```
This file contains average gene expression values for K=8 cell types. Run the following command to get K cell-type mixtures from the ST data X:

```{r, annotate with sc}
K = ncol(sc_ref) # number of cell types
result = annotateWithCorrelations(sc_ref, K, W, H)
H_sc = result$h
W_sc = result$w
```

## Adam: Please add normalization code here as well
```{r, normalize2}
# normalize
W_norm <- W
for(i in 1:nrow(W)){
    W_norm[i,]=W[i,]/sum(W[i,])
}
W_sc_norm <- W_sc
for(i in 1:nrow(W_sc)){
    W_sc_norm[i,]=W_sc[i,]/sum(W_sc[i,])
}
```

<!-- Adam: this function should also give mapping maatrix visualization -->
We assign components to the cell type it has maximum correlation with as shown in figure above. Finally, cell-type proportions for each spot in the ST data (X) are stored in H_est of order K x S.
You can verify that the mappings from both methods are largely similar.

```{r, visualize correlation matrix between the sc reference and W}
correlations = stats::cor(sc_ref, W_norm)
corrplot::corrplot(correlations, 
         is.corr = F, 
         mar=c(0,0,1,0), 
         col = colorRampPalette(c("white", "deepskyblue", "blue4"))(100), 
         main="Correlation-based Mapping Matrix")
```

# Cell-type Annotation via known marker genes

Here, we will annotate using known marker genes for cell types. Load the marker gene list:

```{r, marker}
marker_ref = vignetteColonData$marker_ref
```

This file contains the list of marker genes for K = 8 cell types. Run the following command to get K cell-type mixtures from the ST data X:

```{r, annotate with marker}
K = length(marker_ref) # number of cell types
result = retrofit::annotateWithMarkers(marker_ref, K, W, H)
H_mark = result$h
W_mark = result$w
```

## Adam: Please add normalization code here
```{r, normalize2}
# normalize
H_est <- H_mark
for(i in 1:ncol(H_mark)){
    H_est[,i]=H_mark[,i]/sum(H_mark[,i])
}
W_est <- W_mark
for(i in 1:nrow(W_mark)){
    W_est[i,]=W_mark[i,]/sum(W_mark[i,])
}
```

## Adam: this function should also give mapping maatrix visualization 
```{r, visualize correlation matrix between the marker reference and W}
# correlations = stats::cor(sc_ref_w, W)
# ggcorrplot::ggcorrplot(correlations)
```

We assign components to the cell type it has maximum average marker expression in, as shown in figure above. Finally, cell-type proportions for each spot in the ST data (X) are stored in H_est of order K x S.


# Results and visualization

Hereon, we will be reproducing some of the analysis from the paper using marker-based mapping results.

Figure 4A: Proportion of cell-types in this tissue:

```{r}
rowSums(H_est)/ncol(H_est)
```

Figure 4C: 

Load the coordinates for the spots in the tissue and the tissue image:
```{r}
coords=vignetteColonData$colon_a3_coords
img <- png::readPNG(RCurl::getURLContent("https://user-images.githubusercontent.com/90921267/210159136-96b56551-f414-4b0b-921e-98f05a98c8bc.png"))
t <- grid::rasterGrob(img, width=ggplot2::unit(1,"npc"), height=ggplot2::unit(1,"npc"), interpolate = FALSE)
# img <- png::readPNG(RCurl::getURLContent("https://user-images.githubusercontent.com/90921267/210296599-83b1b839-4680-495e-81db-be172ae41d45.png"))
# t <- grid::rasterGrob(img)
```

Find the cell-type with maximum proportion in each spot and plot:
```{r}
dat=as.data.frame(cbind(coords,t(H_est)))
colnames(dat)=c(colnames(coords),rownames(H_est))

df=dat[,-c(1:3)]
df$CellType=NA
for(i in 1:nrow(df)){
  df$CellType[i]=names(which.max(df[i,-c(1:2)]))
}
df$CellType=factor(df$CellType,levels=c("Epithelial", "Immune", "Myo.Meso","Muscle",
                                        "Neural", "Endothelium", "Pericytes","Fibroblasts"))

ggplot2::ggplot(df, ggplot2::aes(x=imagerow, y=imagecol))  +
  # ggplot2::annotation_custom(t, 700, 6400, 990, 6550) +
  ggplot2::geom_point(size=1.8, shape=21, ggplot2::aes(fill=CellType))+
  ggplot2::scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9","#009E73", "#CC79A7",
                             "#0072B2", "#D55E00" ,"#F0E442"),
                    labels=c("Epithelial", "Immune","Myofibroblasts/\nMesothelium",
                             "Muscle","Neural","Endothelium",
                             "Pericytes","Fibroblasts"))+
  ggplot2::xlab("") + ggplot2::ylab("") + ggplot2::theme_classic()+
  ggplot2::theme(legend.position = "none",
                 axis.line = ggplot2::element_blank(),
                 plot.margin = ggplot2::margin(-0.2, -0.2, 0.2, -1, "cm"), 
                 axis.text.x = ggplot2::element_blank(), axis.text.y = ggplot2::element_blank(),
                 axis.ticks.x = ggplot2::element_blank(),  axis.ticks.y = ggplot2::element_blank())                   
```

Figure 4D and 4E:

Plot the spatial expression for Epithelial markers:
```{r, fig.dim = c(4.5, 4.5*(456/480))}
df = as.data.frame(cbind(coords, t(X[marker_ref$Epithelial,]))) 
df$metagene = unname(rowSums(df[,-c(1:5)])) 
df$metagene[df$metagene>quantile(df$metagene,0.95)]=quantile(df$metagene,0.95)

ggplot2::ggplot(df, ggplot2::aes(x=imagerow, y=imagecol))  +
  # ggplot2::annotation_custom(t, 700, 6400, 990, 6550) +
  ggplot2::geom_point(size=1.5, ggplot2::aes(color=metagene))+
  ggplot2::scale_color_gradientn(name="Epithelial Markers ",
                                 colors=colorspace::adjust_transparency("grey20",
                                                                        alpha = c(0,0.5,1)),
                                 n.breaks=3) +
  ggplot2::xlab("") + ggplot2::ylab("") +
  ggplot2::theme_classic()+
  ggplot2::theme(legend.key.height = ggplot2::unit(0.2, 'cm'), #change legend key size
                 legend.key.width = ggplot2::unit(0.8, 'cm'),
                 legend.title = ggplot2::element_text(size=10),
                 legend.position = "bottom",
                 legend.box.margin=ggplot2::margin(-20,-10,-5,-10),
                 plot.margin = ggplot2::margin(-0.2, -0.2, 0.2, -1, "cm"),
                 axis.line=ggplot2::element_blank(),
                 legend.text=ggplot2::element_text(size=8),
                 axis.text.x = ggplot2::element_blank(), 
                 axis.text.y = ggplot2::element_blank(),
                 axis.ticks.x=ggplot2::element_blank(),
                 axis.ticks.y=ggplot2::element_blank())
```
Plot the estimated proportions for Epithelial cell-type:

```{r, fig.dim = c(4.5, 4.5*(456/480))}
df=dat[,-c(1:3)]
ggplot2::ggplot(df, ggplot2::aes(x=imagerow, y=imagecol))  +
  # ggplot2::annotation_custom(t, 700, 6400, 990, 6550) +
  ggplot2::geom_point(size=1.5, ggplot2::aes(color=Epithelial))+
  ggplot2::scale_color_gradientn(name="Epithelial ",
                                 colors=colorspace::adjust_transparency("grey20",
                                                                        alpha = c(0,0.5,1)),
                                 n.breaks=3, limits=c(0,1)) +
  ggplot2::xlab("") + ggplot2::ylab("") +
  ggplot2::theme_classic()+
  ggplot2::theme(legend.key.height = ggplot2::unit(0.2, 'cm'), #change legend key size
        legend.key.width = ggplot2::unit(0.8, 'cm'),
        legend.title = ggplot2::element_text(size=10),
        legend.position = "bottom",
        legend.box.margin=ggplot2::margin(-20,-10,-5,-10),
        plot.margin = ggplot2::margin(-0.2, -0.2, 0.2, -1, "cm"),
        axis.line=ggplot2::element_blank(),
        legend.text=ggplot2::element_text(size=8),
        axis.text.x = ggplot2::element_blank(), axis.text.y = ggplot2::element_blank(),
        axis.ticks.x=ggplot2::element_blank(),axis.ticks.y=ggplot2::element_blank())
```
Verify that the spatial pattern exhibited by the epithelial markers is largely similar to that of the estimated proportion for epithelial cells.

Figure 5A:

Visualize the proportion of different cell types and colocalization patterns:

```{r}
## Structure Plots
plotfn <- function(df=NULL)
{
  #reshape to long format
  df$num <- 1:nrow(df)
  df1 <- reshape2::melt(df,id.vars = "num")
  #reversing order for cosmetic reasons
  df1 <- df1[rev(1:nrow(df1)),]
  
  #plot
  p <- ggplot2::ggplot(df1, ggplot2::aes(x=num,y=value,fill=variable))+
    ggplot2::geom_bar(stat="identity",position="fill",width = 1, space = 0)+
    ggplot2::scale_x_continuous(expand = c(0, 0))+
    ggplot2::scale_y_continuous(expand = c(0, 0))+
    ggplot2::labs(x = NULL, y = NULL)+
    ggplot2::theme_grey(base_size=7)+
    ggplot2::theme(legend.text = ggplot2::element_text(size = 8),legend.position = "bottom",
          legend.key.size = ggplot2::unit(0.4, "cm"),
          legend.title = ggplot2::element_blank(),
          axis.ticks = ggplot2::element_blank(),
          axis.text.x = ggplot2::element_blank())+
    ggplot2::scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9", 
                               "#009E73", "#CC79A7", "#0072B2",
                               "#D55E00", "#F0E442"))
  
  p
}

df=as.data.frame(t(H_est[c(2,4,6,5,7,1,8,3),]))
#pick max cluster, match max to cluster
maxval <- apply(df,1,max)
matchval <- vector(length=nrow(df))
for(j in 1:nrow(df)) matchval[j] <- match(maxval[j],df[j,])
#add max and match to df
df_q <- df
df_q$maxval <- maxval
df_q$matchval <- matchval
#order dataframe ascending match and decending max
df_q <- df_q[with(df_q, order(matchval,-maxval)), ]
#remove max and match
df_q$maxval <- NULL
df_q$matchval <- NULL


plotfn(df=df_q)+
  ggplot2::theme(legend.position="bottom")+
  ggplot2::ggtitle("Fetal 19 PCW (C)")+ 
  ggplot2::theme(plot.title = ggplot2::element_text(hjust=0.5,size=9))
```

Figure 5D: 

Plot the proportion of only dominant cell types i.e., cell types with proportion > 0.5 in a spot.

```{r, fig.dim = c(4.5, 4.5*(456/480))}
df=dat
df_new=df[df$Epithelial>0.5 | df$Fibroblasts>0.5 | df$Myo.Meso>0.5 |
            df$Endothelium>0.5 | df$Pericytes>0.5 | df$Neural>0.5 |
            df$Immune>0.5 | df$Muscle>0.5,-c(1:3)]
df_new$CellType=1*(df_new$Epithelial>0.5) + 2*(df_new$Immune>0.5) +
  3*(df_new$Myo.Meso>0.5) + 4*(df_new$Muscle>0.5) + 5*(df_new$Neural>0.5)+
  6*(df_new$Endothelium>0.5) + 7*(df_new$Pericytes>0.5) + 8*(df_new$Fibroblasts>0.5)
df_new$CellType=factor(df_new$CellType)

ggplot2::ggplot(df_new, ggplot2::aes(x=imagerow, y=imagecol))  + 
  # ggplot2::annotation_custom(t, 700, 6400, 990, 6550) + 
  ggplot2::geom_point(size=1.8, shape=21, ggplot2::aes(fill=CellType))+
  ggplot2::scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9",
                             "#009E73", "#CC79A7", "#0072B2", 
                             "#D55E00", "#F0E442"),
                    labels=c("Epithelial","Immune", 
                             "Myofibroblasts/\nMesothelium",
                             "Muscle", "Neural", "Endothelium", 
                             "Pericytes", "Fibroblasts"))+ 
  ggplot2::xlab("") + ggplot2::ylab("") +
  ggplot2::theme_classic()+
  ggplot2::theme(legend.key.height = ggplot2::unit(0.5, 'cm'), #change legend key size
        legend.key.width = ggplot2::unit(1, 'cm'), 
        legend.title = ggplot2::element_blank(),
        legend.position = "none",
        plot.margin = ggplot2::margin(-0.2, -0.2, 0.2, -1, "cm"), 
        axis.line=ggplot2::element_blank(),
        legend.text=ggplot2::element_text(size=9),
        axis.text.x = ggplot2::element_blank(), axis.text.y = ggplot2::element_blank(),
        axis.ticks.x=ggplot2::element_blank(),axis.ticks.y=ggplot2::element_blank())
```

Figure 5E: 

Visualize the spots that contain cell types co-localized with Epithelial cells:

```{r, fig.dim = c(4.5, 4.5*(456/480))}
df=dat
df_new=df[df$Epithelial>0.25,]
df_new=df_new[  df_new$Fibroblasts>0.25 | df_new$Myo.Meso>0.25 |
              df_new$Endothelium>0.25 | df_new$Pericytes>0.25 | df_new$Neural>0.25 |
              df_new$Immune>0.25 | df_new$Muscle>0.25,-c(1:3)]

df_new$CellType=NA
for(i in 1:nrow(df_new)){
  df_new$CellType[i]=names(which.max(df_new[i,-c(1:3,5)]))
}
df_new$CellType=factor(df_new$CellType,c("Immune", "Myo.Meso", "Muscle","Neural", "Endothelium",
                                         "Pericytes","Fibroblasts"))

ggplot2::ggplot(df_new, ggplot2::aes(x=imagerow, y=imagecol))  +
  ggplot2::ylim(min(df$imagecol),max(df$imagecol)) +
  ggplot2::xlim(min(df$imagerow),max(df$imagerow))+
  # ggplot2::annotation_custom(t, 700, 6400, 990, 6550) +
  ggplot2::geom_point(size=1.8, shape=21, ggplot2::aes(fill=CellType))+
  ggplot2::scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9",
                             "#009E73", "#CC79A7", "#0072B2",
                             "#D55E00", "#F0E442"),
                    labels=c("Epithelial","Immune",
                             "Myofibroblasts/\nMesothelium",
                             "Muscle", "Neural", "Endothelium",
                             "Pericytes", "Fibroblasts"))+
  ggplot2::xlab("") + ggplot2::ylab("") +
  ggplot2::theme_classic()+
  ggplot2::theme(legend.key.height = ggplot2::unit(0.5, 'cm'), #change legend key size
        legend.key.width = ggplot2::unit(1, 'cm'),
        legend.title = ggplot2::element_blank(),
        legend.position = "none",
        plot.margin = ggplot2::margin(-0.2, -0.2, 0.2, -1, "cm"),
        axis.line = ggplot2::element_blank(),
        axis.text.x = ggplot2::element_blank(), axis.text.y = ggplot2::element_blank(),
        axis.ticks.x=ggplot2::element_blank(),axis.ticks.y=ggplot2::element_blank())
```

Figure 6B:

We found cell-type specific genes using RETROFIT estimates. We filtered genes using the following steps:\

- Common genes in the fetal and adult stages
- Entropy < 1.5 for all replicates
- Gini > 0.85 for all replicates
- Estimated gene expression > 40 for all replicates
- Cell type specificity consistent across replicates 

Now read the relevant genes for this tissue:
```{r}
genes=vignetteColonData$fetal12_genes
```

Plot the normalized gene expression for these genes estimated by RETROFIT:
```{r}
df=as.matrix(W_est[genes,])
df=df[order(row.names(df)),]

dat=data.frame(celltype=factor(c(rep("Epithelial",nrow(df)),rep("Fibroblasts",nrow(df)),
                                 rep("Myofibroblasts\nMesothelium",nrow(df)),rep("Endothelium",nrow(df)),
                                 rep("Pericytes",nrow(df)),rep("Neural",nrow(df)),
                                 rep("Immune",nrow(df)),rep("Muscle",nrow(df))),
                               levels=c("Epithelial","Immune","Fibroblasts","Endothelium",
                                        "Neural","Pericytes", "Myofibroblasts\nMesothelium","Muscle")),
               genes=factor(genes,levels=genes),
               val=c(df[,"Epithelial"],df[,"Fibroblasts"],
                     df[,"Myo.Meso"],df[,"Endothelium"],
                     df[,"Pericytes"],df[,"Neural"],
                     df[,"Immune"],df[,"Muscle"]))

ggplot2::ggplot(dat, ggplot2::aes(celltype, genes, fill= val)) +
  ggplot2::geom_tile() + ggplot2::theme_classic() +
  ggplot2::scale_fill_gradientn(colors = pals::brewer.reds(20)[2:20],
                                name="Normalized\nGene Expression   ",
                                limit=c(0,1), na.value="grey87"
  )+
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1,size=8),
                 legend.key.height = ggplot2::unit(0.2, 'cm'),
                 legend.key.width = ggplot2::unit(1, 'cm'),
                 legend.title = ggplot2::element_text(size=7),
                 legend.justification="center",
                 legend.box.margin=ggplot2::margin(-20,-10,-5,-10),
                 legend.text=ggplot2::element_text(size=7)) +
  ggplot2::xlab('')+ ggplot2::ylab('')+
  ggplot2::theme(legend.position="bottom")
```


# Session information

```{r}
sessionInfo()
```



